### DISCUSSION TITLE
Breaking change "cf" in v3.4.0: clarification needed

### DISCUSSION THREAD

**Discussion by cs-1** [netbox-community/netbox#11195] [Q&A] [answered]
Hi all,

I have a quick question regarding the changelog item

- Within the Python API, the cf property on an object with custom fields now returns deserialized values. For example, a custom field referencing an object will return the object instance rather than its numeric ID. To access the raw serialized values, reference the object's custom_field_data attribute instead.

"Python API" means that this concerns code written within the Python framework (e.g. custom scripts and reports as well as plugins) and not the restful API (where custom fields can be accessed via e. g. `interface["custom_fields"]`), is that correct?

And all Python API code (e.g. Reports) that use `custom_field_data` is not affected?

I'm terribly sorry for the stupid questions, I'm not a native speaker and have only recently started coding Reports (loving it, BTW, the source code of NetBox is easy to read and it's fun to explore things in nbshell).

**Comment by jeremystretch** [comment, marked as answer]
> "Python API" means that this concerns code written within the Python framework (e.g. custom scripts and reports as well as plugins) and not the restful API (where custom fields can be accessed via e. g. interface["custom_fields"]), is that correct?

Correct. Here's an example:

```python
>>> site = Site.objects.first()
>>> site.cf['date1']
datetime.date(2022, 12, 13)
```

> And all Python API code (e.g. Reports) that use custom_field_data is not affected?

Correct. No changes have been made to the serialized data stored in the database.

**Reply by ziggekatten** [reply]
Clarification makes us more confused. We jus realized that all our x.cf['key'] on our custom scripts silently neither create or update data after upgrade to 3.4.x. Changing to x.custom_field_data[key] = 'data'  works however. 

Is this according to what you mean by the breaking change, that x.cf[] is not to be used anymore?

**Reply by jeremystretch** [reply]
The `cf` alias was never intended to be used for _assigning_ data; it was introduced as a shorthand for accessing custom field data primarily in templates.

**Reply by ziggekatten** [reply]
> The `cf` alias was never intended to be used for _assigning_ data; it was introduced as a shorthand for accessing custom field data primarily in templates.

Ohhh...bummer...

**Comment by cs-1** [comment]
Thanks for your quick reply, Jeremy, it's highly appreciated. That answers my question.
